backends/dialogs/win32/win32-dialogs.cpp

```diff
@@ -0,0 +1,208 @@
+#include "config.h"
+#endif
+
+#if defined(WIN32) && defined(USE_SYSDIALOGS)
+
+// HACK: To get __MINGW64_VERSION_foo defines we need to manually include
+// _mingw.h in this file because we do not include any system headers at this
+// point on purpose. The defines are required to detect whether this is a
+// classic MinGW toolchain or a MinGW-w64 based one.
+#if defined(__MINGW32__)
+#include <_mingw.h>
+#endif
+
+// Needed for dialog functions
+// HACK: MinGW-w64 based toolchains include the symbols we require in their
+// headers. The 32 bit incarnation only defines __MINGW32__. This leads to
+// build breakage due to clashes with our compat header. Luckily MinGW-w64
+// based toolchains define __MINGW64_VERSION_foo macros inside _mingw.h,
+// which is included from all system headers. Thus we abuse that to detect
+// them.
+#if defined(__GNUC__) && defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR)
+#include "backends/dialogs/win32/mingw-compat.h"
+#else
+	// We use functionality introduced with Vista in this file.
+	// To assure that including the respective system headers gives us all
+	// required definitions we set Vista as minimum version we target.
+	// See: https://msdn.microsoft.com/en-us/library/windows/desktop/aa383745%28v=vs.85%29.aspx#macros_for_conditional_declarations
+#include <sdkddkver.h>
+#undef _WIN32_WINNT
+#define _WIN32_WINNT _WIN32_WINNT_VISTA
+
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#if defined(ARRAYSIZE)
+#undef ARRAYSIZE
+#endif
+#endif
+
+#include <shlobj.h>
+
+#include "common/scummsys.h"
+
+#include "backends/dialogs/win32/win32-dialogs.h"
+#include "backends/platform/sdl/win32/win32_wrapper.h"
+#include "backends/platform/sdl/win32/win32-window.h"
+
+#include "common/config-manager.h"
+#include "common/system.h"
+#include "common/events.h"
+#include "common/translation.h"
+
+Win32DialogManager::Win32DialogManager(SdlWindow_Win32 *window) : _window(window) {
+	CoInitialize(NULL);
+}
+
+Win32DialogManager::~Win32DialogManager() {
+	CoUninitialize();
+}
+
+// Wrapper for old Windows versions
+HRESULT winCreateItemFromParsingName(PCWSTR pszPath, IBindCtx *pbc, REFIID riid, void **ppv) {
+	typedef HRESULT(WINAPI *SHFunc)(PCWSTR, IBindCtx *, REFIID, void **);
+
+	SHFunc func = (SHFunc)GetProcAddress(GetModuleHandle(TEXT("shell32.dll")), "SHCreateItemFromParsingName");
+	if (func == NULL)
+		return E_NOTIMPL;
+
+	return func(pszPath, pbc, riid, ppv);
+}
+
+HRESULT getShellPath(IShellItem *item, Common::String &path) {
+	LPWSTR name = NULL;
+	HRESULT hr = item->GetDisplayName(SIGDN_FILESYSPATH, &name);
+	if (SUCCEEDED(hr)) {
+		char *str = Win32::unicodeToAnsi(name);
+		path = Common::String(str);
+		CoTaskMemFree(name);
+		delete[] str;
+	}
+	return hr;
+}
+
+Common::DialogManager::DialogResult Win32DialogManager::showFileBrowser(const char *title, Common::FSNode &choice, bool isDirBrowser) {
+	DialogResult result = kDialogError;
+
+	// Do nothing if not running on Windows Vista or later
```

Comment:
Wouldn't it be better to just disable the feature altogether for earlier Windows versions? (i.e. XP or earlier)

Reply:
I wanted to keep the option open for devs to add dialogs with different compatibility levels. Returning an error effectively disables the feature, as it'll fall back to the GUI dialog.